### DISCUSSION TITLE
How to access System.__ComObject wrapper

### DISCUSSION THREAD

**Discussion by faljse** [sebastienros/jint#1418] [Q&A] [answered]
C# creates a com wrapper around an api i need to use, so i am unable to access any method or
field of this object.
Objects are of type {System.__ComObject}.

Is there a way to interop with this api?

**Comment by lahma** [comment, marked as answer]
You probably need to define you own type inheriting from `ObjectInstance` and add methods and properties you need, them calling your CLR methods that handle the invoke.

**Reply by faljse** [reply]
Thx, i can now access the methods from my wrapper/proxy class; however Object.keys returns an empty array on my type.:
```
engine.SetValue("au", new JHAutomaticProxy(engine, au));
```
[..]
```
public class JHAutomaticProxy: ObjectInstance
{
  private JHAutomatic au;
  public JHAutomaticProxy(Engine engine, JHAutomatic au): base(engine)
  {
      this.au = au;
  }
  
  public void StartProgram(string bstrProgramName)
  {
      au.StartProgram(bstrProgramName);
  }
```

```
jint> Object.keys(au);
[]
```

**Reply by lahma** [reply]
You need to register your properties. See the `Initialize` [here for example](https://github.com/sebastienros/jint/blob/main/Jint/Native/Object/ObjectConstructor.cs).

**Reply by faljse** [reply]
Got it working for a single property.
This is getting more complex than expected; since basic interop is working fine (calling methods).
Can i just life with empty keys(), and skip register properties, or will this cause issues later?

**Reply by lahma** [reply]
You can check how ObjectWrapper in Jint code base dynamically searches and adds members and see if you can adapt it.

Your mileage may vary with the properties not being present, like you've seen they won't be visible to enumeration for example.
